testsupport/user_request.go

```diff
@@ -0,0 +1,265 @@
+}
+
+func (r *userRequest) Username(username string) UserRequest {
+	r.username = &username
+	return r
+}
+
+func (r *userRequest) Email(email string) UserRequest {
+	r.email = &email
+	return r
+}
+
+func (r *userRequest) Resources() (*toolchainv1alpha1.UserSignup, *toolchainv1alpha1.MasterUserRecord) {
+	return r.userSignup, r.mur
+}
+
+func (r *userRequest) EnsureMUR() UserRequest {
+	r.ensureMUR = true
+	return r
+}
+
+func (r *userRequest) ManuallyApprove() UserRequest {
+	r.manuallyApprove = true
+	return r
+}
+
+func (r *userRequest) RequireConditions(conditions ...toolchainv1alpha1.Condition) UserRequest {
+	r.conditions = conditions
+	return r
+}
+
+func (r *userRequest) VerificationRequired() UserRequest {
+	r.verificationRequired = true
+	return r
+}
+
+func (r *userRequest) TargetCluster(targetCluster *wait.MemberAwaitility) UserRequest {
+	r.targetCluster = targetCluster
+	return r
+}
+
+func (r *userRequest) RequireHTTPStatus(httpStatus int) UserRequest {
+	r.requiredHTTPStatus = httpStatus
+	return r
+}
+
+func (r *userRequest) Execute() UserRequest {
+	WaitUntilBaseNSTemplateTierIsUpdated(r.t, r.hostAwait)
+
+	var username string
+	if r.username != nil {
+		username = *r.username
+	} else {
+		username = fmt.Sprintf("testuser-%s", uuid.Must(uuid.NewV4()).String())
+	}
+
+	// Create a token and identity to sign up with
+	userIdentity := &authsupport.Identity{
+		ID:       uuid.Must(uuid.NewV4()),
+		Username: username,
+	}
+
+	var email string
+	if r.email != nil {
+		email = *r.email
+	} else {
+		email = fmt.Sprintf("%s@test.com", username)
+	}
+
+	emailClaim0 := authsupport.WithEmailClaim(email)
+	token0, err := authsupport.GenerateSignedE2ETestToken(*userIdentity, emailClaim0)
+	require.NoError(r.t, err)
+
+	// Call the signup endpoint
+	invokeEndpoint(r.t, "POST", r.hostAwait.RegistrationServiceURL+"/api/v1/signup",
+		token0, "", r.requiredHTTPStatus)
+
+	// Wait for the UserSignup to be created
+	userSignup, err := r.hostAwait.WaitForUserSignup(userIdentity.ID.String())
+	require.NoError(r.t, err)
+
+	if r.manuallyApprove || r.targetCluster != nil || (r.verificationRequired != states.VerificationRequired(userSignup)) {
+		doUpdate := func(instance *toolchainv1alpha1.UserSignup) error {
+			// We set the VerificationRequired state first, because if manuallyApprove is also set then it will
+			// reset the VerificationRequired state to false.
+			if r.verificationRequired != states.VerificationRequired(instance) {
+				states.SetVerificationRequired(userSignup, r.verificationRequired)
+			}
+
+			if r.manuallyApprove {
+				states.SetApproved(instance, r.manuallyApprove)
+			}
+			if r.targetCluster != nil {
+				instance.Spec.TargetCluster = r.targetCluster.ClusterName
+			}
+			return r.hostAwait.FrameworkClient.Update(context.TODO(), instance)
+		}
+
+		err = doUpdate(userSignup)
+		if err != nil {
+			// Under certain conditions, the controller has already updated the UserSignup so it's stale.  This block
```

Comment:
Shouldn't we check that the error is 409 (conflict)? And fail right away if it is not?
And we actually already have `UpdateUserSignupSpec()` function which you can use to update the UserSignup and which will deal with 409 conflicts: https://github.com/codeready-toolchain/toolchain-e2e/blob/master/testsupport/wait/host.go#L147

Reply:
Ah that's a useful function... I've updated to use that instead.